### PR TITLE
RPC Documents Consolidation

### PR DESCRIPTION
**Pull request by ShahanaFarooqui**
This PR will generate `doc/lightning-*.md` file contents for rpc commands. It will get the information from consolidated `doc/schemas/lightning-*.json` and use it to generate markdown for RPC commands.

Changelog-Changed: Documentation: great documentation rewrite, all reference pages now generated from the fully-tested JSON schemas and include examples.

### CODE REVIEW

**Comment by cdecker** [comment]
@ShahanaFarooqui, this is tagged for version v24.02. I plan to call a feature freeze on the 29th, and given that this PR is still drafted I wanted to ask whether you are planning on finalizing this by then, or if more time is needed.

**Comment by ShahanaFarooqui** [comment]
@cdecker I need some help in fixing the CI and I am hoping to be able to do that around Jan 26th. I will update the milestone if it is not ready by then.

**Comment by tonyaldon** [comment]
Hey @ShahanaFarooqui, in `doc/schemas/lightning-openchannel_abort.json` file you are using `\\_` and maybe this is not what you want (but I'm not sure):

    line  5:   "rpc": "openchannel\\_abort",
    line 63:     "lightning-openchannel\\_init(7)",
    line 64:     "lightning-openchannel\\_update(7)",
    line 65:     "lightning-openchannel\\_signed(7)",
    line 66:     "lightning-openchannel\\_bump(7)",
    line 67:     "lightning-fundchannel\\_start(7)",
    line 68:     "lightning-fundchannel\\_complete(7)",

**Comment by tonyaldon** [comment]
Hey @ShahanaFarooqui, in `doc/schemas/lightning-listdatastore.json` and `doc/schemas/lightning-help.json` file, you are using `\n` and `\n\t` and I'm not sure this is what you want to do though you don't do this in the other schemas:

File lightning-listdatastore.json:

    line 18:  "all immediate children of the *key* (or root children) are returned.\n        Using the first element of the key as the plugin name (e.g. `[ 'summary' ]`) is recommended.\n        An array of values to form a hierarchy (though a single value is treated as a one-element array)"

File lightning-help.json:

    line 61:           "verbose": "check command_to_check [parameters...]\n"
    line 41:           "verbose": "dev rhash {secret}\n\tShow SHA256 of {secret}\ndev crash\n\tCrash lightningd by calling fatal()\ndev slowcmd {msec}\n\tTorture test for slow commands, optional {msec}\n"
    line 37:           "verbose": "help [command]\nWithout [command]:\n  Outputs an array of objects with 'command' and 'description'\nWith [command]:\n  Give a single object containing 'verbose', which completely describes\n  the command inputs and outputs."
    line 43:           "verbose": "listconfigs [config]\nOutputs an object, with each field a config options\n(Option names which start with # are comments)\nWith [config], object only has that field"
    line 30:           "verbose": "Usage :\nplugin start /path/to/a/plugin\n\tadds a new plugin to Core Lightning\nplugin stop plugin_name\n\tstops an already registered plugin\nplugin startdir /path/to/a/plugin_dir/\n\tadds a new plugin directory\nplugin rescan\n\tloads not-already-loaded plugins from the default plugins dir\nplugin list\n\tlists all active plugins\n\n"

**Comment by ShahanaFarooqui** [comment]
@tonyaldon Thanks for pointing these out. Cleaned them up with some other updates.

File: plugins/Makefile [hunk 1]
```diff
@@ -225,16 +225,26 @@ plugins/recover:  common/gossmap.o common/fp16.o $(PLUGIN_RECOVER_OBJS) $(PLUGIN
...
-# We can't simply *remove* fields, since the extra comma left over can
-# make invalid JSON. Grr!
-# But these simple removals drop us from 100k to 29k.
+SQL_LISTRPCS_SCHEMAS := $(foreach l,$(SQL_LISTRPCS),doc/schemas/lightning-$l.json)
+
+# RES_JSON is generated by:
+# 1: deleting description array (used for field details) but keeping description keys
+# 2: deleting additionalProperties, required, enum, maxLength, minLength, pre_return_value_notes, post_return_value_notes, anyOf, oneOf, hidden, untyped, default, comment, added, maximum, minimum, pattern, format
+# (We only need [field].type, [field].name, [field].deprecated, [field].allOf, [field].items.allOf and allOf from response.properties)
+# 3: deleting empty objects from .response.properties
```
Comment (rustyrussell) [inline plugins/Makefile:234]: Nice!  This actually reduces the header file from 44081 bytes to 18477!

File: contrib/pyln-testing/pyln/testing/fixtures.py [hunk 2]
```diff
@@ -432,36 +432,28 @@ def is_msat_or_any_or_currency(checker, instance):
...
-            base = fname.rpartition('.request')[0]
-            is_request = True
-            index = 0
-        else:
-            continue
-        if base not in schemas:
+        if fname.startswith('lightning-') and fname.endswith('.json'):
+            base = fname.replace('lightning-', '').replace('.json', '')
+            # Request is 0 and Response is 1
             schemas[base] = [None, None]
```
Comment (rustyrussell) [inline contrib/pyln-testing/pyln/testing/fixtures.py:455]: This line can also be deleted?

File: doc/lightning-addgossip.7.md [hunk 3]
```diff
@@ -1,31 +1,54 @@
...
 On success, an empty object is returned.
 
-[comment]: # (GENERATE-FROM-SCHEMA-END)
+EXAMPLE JSON RESPONSE
+---------------------
+
+```json
+{}
+{}
+```
```
Comment (rustyrussell) [inline doc/lightning-addgossip.7.md]: Does this really return two empty objects?

File: doc/lightning-addpsbtoutput.7.md [hunk 4]
```diff
@@ -1,46 +1,104 @@
...
-by adding a single output of amount *satoshi*.
+Command *added* in v23.11.
 
-This is used to receive funds into the on-chain wallet interactively
-using PSBTs.
+`addpsbtoutput` is a low-level RPC command which creates or modifies a PSBT by adding a single output of amount *satoshi*.
+
+This is used to receive funds into the on-chain wallet interactively using PSBTs.
+
+- **satoshi** (msat): The satoshi value of the output. It can be a whole number, a whole number ending in *sat*, a whole number ending in *000msat*, or a number with 1 to 8 decimal places ending in *btc*.
```
Comment (rustyrussell) [inline doc/lightning-addpsbtoutput.7.md]: This is actually type "sat", which matters if it's a raw number... (I just checked the code!)

File: doc/lightning-addgossip.7.md [hunk 5]
```diff
@@ -1,31 +1,54 @@
 lightning-addgossip -- Command for injecting a gossip message (low-level)
-===============================================================
+=========================================================================
 
 SYNOPSIS
```
Comment (rustyrussell) [inline doc/lightning-addgossip.7.md]: This commit needs a Changelog message!

Changelog-Added: doc: Major command reference rework, including examples and documentation generated from the fully-tested JSON schemas.

File: doc/contribute-to-core-lightning/code-generation.md [hunk 6]
```diff
@@ -13,19 +13,11 @@ The CLN project has a multitude of interfaces, most of which are generated from
...
 
-- We describe the JSON-RPC using [JSON Schema](https://json-schema.org/) in the [`doc/schemas`](https://github.com/ElementsProject/lightning/tree/master/doc/schemas) directory. Each method has a `.request.json` for the request message, and a `.schema.json` for the response (the mismatch is historical and will eventually be addressed). During tests the `pytest` target will verify responses, however the JSON-RPC methods are _not_ generated (yet?). We do generate various client stubs for languages, using the `msggen`][msggen] tool. More on the generated stubs and utilities below.
+- We describe the JSON-RPC using [JSON Schema](https://json-schema.org/) in the [`doc/schemas`](https://github.com/ElementsProject/lightning/tree/master/doc/schemas) directory. Each method has a `lightning-*.json` for request and response. During tests the `pytest` target will verify responses, however the JSON-RPC methods are _not_ generated (yet?). We do generate various client stubs for languages, using the `msggen`[msggen] tool. More on the generated stubs and utilities below.
```
Comment (rustyrussell) [inline doc/contribute-to-core-lightning/code-generation.md:16]: This needs updating: we *do* check both request and response now!

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md [hunk 7]
```diff
@@ -38,55 +38,63 @@ The `getmanifest` method is required for all plugins and will be called on start
...
+		{
+			"method": "mycustomnotification"
+		}
+	],
+	"custommessages": [
+		11008,
+		11010
+	],
+	"nonnumericids": true,
+	"dynamic": true
```
Comment (rustyrussell) [inline doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md:97]: Formatting here is weird, maybe pass through jq?

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md [hunk 8]
```diff
@@ -129,38 +137,38 @@ Nota bene: if a `flag` type option is not set, it will not appear in the options
...
+			"description": "Port to use to connect to 3rd-party service"
+		},
+		{
+			"name": "number",
+			"type": "int",
+			"default": 0,
+			"description": "Another number to add",
+			"multi": true
+		}
+	]
```
Comment (rustyrussell) [inline doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md:171]: Here too

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md [hunk 9]
```diff
@@ -175,7 +183,7 @@ When forwarding a custom notification `lightningd` will wrap the payload of the
   "method": "mycustomnotification",
   "params": {
     "key": "value",
-	"message": "Hello fellow plugin!"
+	  "message": "Hello fellow plugin!"
```
Comment (rustyrussell) [inline doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md:186]: Here too

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md [hunk 10]
```diff
@@ -205,29 +213,31 @@ The `init` method is required so that `lightningd` can pass back the filled comm
...
+			"invoice": "028200"
+		},
+		"proxy": {
+			"type": "ipv4",
+			"address": "127.0.0.1",
+			"port": 9050
+		},
+		"torv3-enabled": true,
+		"always_use_proxy": false
+	}
```
Comment (rustyrussell) [inline doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md:240]: Here too

File: contrib/msggen/msggen/schema.json [hunk 11]
```diff
@@ -29534,272 +29534,6 @@
       "Main web site: <https://github.com/ElementsProject/lightning>"
     ]
   },
-  "lightning-sql.json": {
-    "$schema": "../rpc-schema-draft.json",
-    "type": "object",
```
Comment (rustyrussell) [inline contrib/msggen/msggen/schema.json:29539]: Mayube fold last commit into previous one (or wherever it is needed?)

**Review by rustyrussell — APPROVED** [review]
Amazing work!

File: doc/lightning-addgossip.7.md (hunk 5 above)
Comment (rustyrussell) [reply doc/lightning-addgossip.7.md]: Nope, I see you did this later, and even better :orange_heart:

File: plugins/Makefile (hunk 1 above)
Comment (ShahanaFarooqui) [reply plugins/Makefile:234]: Added it into the comments too.

File: contrib/pyln-testing/pyln/testing/fixtures.py (hunk 2 above)
Comment (ShahanaFarooqui) [reply contrib/pyln-testing/pyln/testing/fixtures.py:455]: These lines should not be deleted to ensure that the `base` is getting the correct rpc method name and only rpc files in the schema are being processed.

File: doc/lightning-addgossip.7.md (hunk 3 above)
Comment (ShahanaFarooqui) [reply doc/lightning-addgossip.7.md]: Addgossip and most of the other schemas, have two example json requests and their corresponding two json responses. So one of the empty object was returned from first request:

```
   {
      "id": "example:addgossip#1",
      "method": "addgossip",
      "params": {
        "message": "010078c3314666731e339c0b8434f7824797a084ed7ca3655991a672da068e2c44cb53b57b53a296c133bc879109a8931dc31e6913a4bda3d58559b99b95663e6d52775579447ef5526300e1bb89bc6af8557aa1c3810a91814eafad6d103f43182e17b16644cb38c1d58a8edd094303959a9f1f9d42ff6c32a21f9c118531f512c8679cabaccc6e39dbd95a4dac90e75a258893c3aa3f733d1b8890174d5ddea8003cadffe557773c54d2c07ca1d535c4bf85885f879ae466c16a516e8ffcfec1740e3f5c98ca9ce13f452e867befef5517f306ed6aa5119b79059bcc6f68f329986b665d16de7bc7df64e3537504c91eeabe0e59d3a2b68e4216ead2b0f6e3ef7c000006226e46111a0b59caaf126043eb5bbf28c34f3a5e332a1fc7b2b73cf188910f0000670000010000022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d590266e4598d1d3c415f572a8488830b60f7e744ed9235eb0b1ba93283b315c0351802e3bd38009866c9da8ec4aa99cc4ea9c6c0dd46df15c61ef0ce1f271291714e5702324266de8403b3ab157a09f1f784d587af61831c998c151bcc21bb74c2b2314b"
      }
    }
```

and then the second is returned from the next example request:
```
    {
      "id": "example:addgossip#2",
      "method": "addgossip",
      "params": {
        "message": "0102420526c8eb62ec6999bbee5f1de4841cab734374ec642b7deeb0259e76220bf82e97a241c907d5ff52019655f7f9a614c285bb35690f3a1a2b928d7b2349a79e06226e46111a0b59caaf126043eb5bbf28c34f3a5e332a1fc7b2b73cf188910f000067000001000065b32a0e010100060000000000000000000000010000000a000000003b023380"
      }
    }
```

File: doc/lightning-addpsbtoutput.7.md (hunk 4 above)
Comment (ShahanaFarooqui) [reply doc/lightning-addpsbtoutput.7.md]: Thanks, I missed it. Corrected now.

File: doc/lightning-addgossip.7.md (hunk 5 above)
Comment (ShahanaFarooqui) [reply doc/lightning-addgossip.7.md]: Updated the changelog on this page too.

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md (hunk 7 above)
Comment (ShahanaFarooqui) [reply doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md]: Done.

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md (hunk 9 above)
Comment (ShahanaFarooqui) [reply doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md]: Done.

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md (hunk 10 above)
Comment (ShahanaFarooqui) [reply doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md]: Done.

File: contrib/msggen/msggen/schema.json (hunk 11 above)
Comment (ShahanaFarooqui) [reply contrib/msggen/msggen/schema.json]: I squashed two commits related to lightning-sql into one commit `docs: Updates script and fixed test to generate lightning-sql`. I hope that fixed this one too. If not, could you please specify which commit you want folded?

File: doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md (hunk 8 above)
Comment (ShahanaFarooqui) [reply doc/developers-guide/plugin-development/a-day-in-the-life-of-a-plugin.md]: Done.

File: contrib/pyln-testing/pyln/testing/fixtures.py (hunk 2 above)
Comment (rustyrussell) [reply contrib/pyln-testing/pyln/testing/fixtures.py:455]: I mean, the code now looks like:

```
            schemas[base] = [None, None]
            schemas[base] = _load_schema(os.path.join('doc/schemas', fname))
```

The first line of these can be deleted!

**Review by rustyrussell — CHANGES_REQUESTED** [review]

File: cln-rpc/src/model.rs [hunk 12]
```diff
@@ -1928,7 +1982,6 @@ pub mod requests {
 	}
 	#[derive(Clone, Debug, Deserialize, Serialize)]
 	pub struct OfferRequest {
-	    pub amount: String,
```
Comment (daywalker90) [inline cln-rpc/src/model.rs:1931]: Why is the amount just removed here?

File: contrib/pyln-testing/pyln/testing/fixtures.py (hunk 2 above)
Comment (ShahanaFarooqui) [reply contrib/pyln-testing/pyln/testing/fixtures.py]: Got it :). Removed `schemas[base] = [None, None]`.

File: cln-rpc/src/model.rs (hunk 12 above)
Comment (ShahanaFarooqui) [reply cln-rpc/src/model.rs:1931]: Thanks for pointing it out. cln-rpc/src/model.rs file is auto generated. Previously, the amount was defined as "string" and now it is updated to "oneOf": [{"type": "msat_or_any"}, {"type": "currency"}]. Looks like model generation is unable to generate from `oneOf` type.

@cdecker Could you please help in generating the model.rs correctly? Or I can change it back to "string" type.

Comment (ShahanaFarooqui) [reply cln-rpc/src/model.rs]: Ignore my above message please, found the fix by adding Offer.amount in overrides array.

**Pull request merged**